File: code/game/Command.cpp
#include "Command.hpp"
#include "CommandFactory.hpp"
#include "Engine.hpp"
#include "BaalExceptions.hpp"
#include "Interface.hpp"
#include "Spell.hpp"
#include "SpellFactory.hpp"
#include "World.hpp"
#include "Player.hpp"
#include "Drawable.hpp"

#include <ctime>
#include <sstream>
#include <memory>

using namespace baal;

namespace {

///////////////////////////////////////////////////////////////////////////////
std::string create_help_str(const Command* command,
                            const std::string& usage)
///////////////////////////////////////////////////////////////////////////////
{
  const CommandFactory& factory = CommandFactory::instance();

  // Get command name
  std::string command_name = factory.name(command);

  // Get command aliases
  std::vector<std::string> aliases;
  factory.aliases(command_name, aliases);

  // Transform aliases into string
  std::string alias_str;
  if (!aliases.empty()) {
    for (std::vector<std::string>::const_iterator
         itr = aliases.begin(); itr != aliases.end(); ++itr) {
      alias_str += *itr + " ";
    }
  }

  // Formulate and return full help string
  std::string help_str = command_name + " " + usage;
  if (!aliases.empty()) {
    help_str += "  Aliases: " + alias_str;
  }
  return help_str;
}

} // empty namespace

///////////////////////////////////////////////////////////////////////////////
void HelpCommand::init(const std::vector<std::string>& args)
///////////////////////////////////////////////////////////////////////////////
{
  RequireUser(args.size() <= 1, "The help command takes at most one argument");

  if (!args.empty()) {
    m_arg = args.front();

    // Verify m_arg is a valid command name
    const CommandFactory& factory = CommandFactory::instance();
    RequireUser(factory.m_cmd_map.find(m_arg) != factory.m_cmd_map.end(),
                "Cannot get help for unknown command " << m_arg);
  }
  else {
    m_arg = "";
  }
}

///////////////////////////////////////////////////////////////////////////////
void HelpCommand::apply(Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  const std::map<std::string, Command*>& cmd_map =
    CommandFactory::instance().m_cmd_map;

  if (m_arg.empty()) {
    std::string help_msg = "List of available commands:\n\n";
    for (std::map<std::string, Command*>::const_iterator
         itr = cmd_map.begin();
         itr != cmd_map.end();
         ++itr) {
      help_msg += itr->second->help(engine) + "\n\n";
    }
    engine.interface().help(help_msg);
  }
  else {
    std::map<std::string, Command*>::const_iterator
      itr = cmd_map.find(m_arg);
    Require(itr != cmd_map.end(), "Command " << m_arg << " missing from map");
    engine.interface().help(itr->second->help(engine));
  }
}

///////////////////////////////////////////////////////////////////////////////
std::string HelpCommand::help(const Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  return create_help_str(this,
"[command]\n"
"  Returns info/syntax help for a command\n"
                         );
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
void EndTurnCommand::init(const std::vector<std::string>& args)
///////////////////////////////////////////////////////////////////////////////
{
  RequireUser(args.empty(), "The end command takes no arguments");
}

///////////////////////////////////////////////////////////////////////////////
void EndTurnCommand::apply(Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  engine.interface().end_turn();
}

///////////////////////////////////////////////////////////////////////////////
std::string EndTurnCommand::help(const Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  return create_help_str(this,
"\n"
"  Ends the current turn\n"
                         );
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
void QuitCommand::init(const std::vector<std::string>& args)
///////////////////////////////////////////////////////////////////////////////
{
  RequireUser(args.empty(), "The quit command takes no arguments");
}

///////////////////////////////////////////////////////////////////////////////
void QuitCommand::apply(Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  engine.interface().end_turn();
  engine.quit();
}

///////////////////////////////////////////////////////////////////////////////
std::string QuitCommand::help(const Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  return create_help_str(this,
"\n"
"  Ends the game\n"
                         );
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
void SaveCommand::init(const std::vector<std::string>& args)
///////////////////////////////////////////////////////////////////////////////
{
  RequireUser(args.size() <= 1, "The save command takes at most one argument");

  if (!args.empty()) {
    m_arg = args.front();
  }
  else {
    // Create a save name based on time/date

    std::ostringstream out;

    // Get current time and convert to tm struct
    time_t curr_time_out = time(NULL);
    struct tm* UTC_time_out = localtime(&curr_time_out);

    // Create save name
    out << "baal_"
        << UTC_time_out->tm_year + 1900 << "-"
        << UTC_time_out->tm_mon + 1 << "-"
        << UTC_time_out->tm_mday << "__"
        << UTC_time_out->tm_hour << ":"
        << UTC_time_out->tm_min << ":"
        << UTC_time_out->tm_sec
        << ".save";

    m_arg = out.str();
  }
}

///////////////////////////////////////////////////////////////////////////////
void SaveCommand::apply(Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  // TODO
}

///////////////////////////////////////////////////////////////////////////////
std::string SaveCommand::help(const Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  return create_help_str(this,
"[save filename]\n"
"  Saves the game; if no name provided, a name based on data/time will be used\n"
                     );
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
void SpellCommand::init(const std::vector<std::string>& args)
///////////////////////////////////////////////////////////////////////////////
{
  RequireUser(args.size() == 3, "The cast command takes 3 arguments");

  // Parse spell name
  m_spell_name = args[0];

  // Parse spell level
  std::istringstream iss(args[1]);
  iss >> m_spell_level;
  RequireUser(!iss.fail(), "Second argument not a valid integer");

  // Parse location
  try {
    m_spell_location = Location(args[2]);
  }
  catch (ProgramError& e) {
    RequireUser(false, "Third argument was not a valid location. " <<
                "Expect <row>,<col> (no spaces)\nError was: " << e.what());
  }
}

///////////////////////////////////////////////////////////////////////////////
void SpellCommand::apply(Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  World&  world  = engine.world();
  Player& player = engine.player();

  // Ensure location is in-bounds
  RequireUser(world.in_bounds(m_spell_location),
              "Location " << m_spell_location << " out of bounds. " <<
              "Max row is: " << world.height() - 1 <<
              ", max col is: " << world.width() - 1);

  // Create the spell. Use an auto_ptr to ensure deletion even if a throw
  // happens. I'd rather use a reference here since spell cannot be NULL, but
  // we need to auto_ptr since verify_cast can throw exceptions.
  std::auto_ptr<const Spell> spell(
    &(SpellFactory::create_spell(m_spell_name,
                                 m_spell_level,
                                 m_spell_location))
                                   );

  // Verify that player can cast this spell (can throw)
  player.verify_cast(*spell);

  // These last two operations need to be atomic, neither should ever throw
  // a user error.

  // Let the player object know that the spell has been cast and to adjust
  // it's state accordingly.
  player.cast(*spell);

  // Apply the spell to the world
  spell->apply(world);
}

///////////////////////////////////////////////////////////////////////////////
std::string SpellCommand::help(const Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  const TalentTree& talents = engine.player().talents();

  std::vector<std::pair<std::string, unsigned> > castable_spells;
  talents.query_all_castable_spells(castable_spells);

  std::string help_str =
"<spell-name> <level> <row>,<col>\n"
"  Casts spell of type <spell-name> and level <level> at location <row>,<col>\n"
"  Castable spells:\n";
  for (std::vector<std::pair<std::string, unsigned> >::const_iterator
       itr = castable_spells.begin(); itr != castable_spells.end(); ++itr) {
    std::string spell_name = itr->first;
    unsigned spell_lvl     = itr->second;
    std::ostringstream out;
    out << "    " << spell_name << " : " << spell_lvl << "\n";
    help_str += out.str();
  }

  return create_help_str(this, help_str);
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
void LearnCommand::init(const std::vector<std::string>& args)
///////////////////////////////////////////////////////////////////////////////
{
  RequireUser(args.size() == 2, "The learn command takes 2 arguments");

  // Parse spell name
  m_spell_name = args[0];

  // Parse spell level
  std::istringstream iss(args[1]);
  iss >> m_spell_level;
  RequireUser(!iss.fail(), "Second argument not a valid integer");
}

///////////////////////////////////////////////////////////////////////////////
void LearnCommand::apply(Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  Player& player = engine.player();
  Location dummy;

  // We need to auto_ptr since learn can throw exceptions.
  std::auto_ptr<const Spell> spell(
    &(SpellFactory::create_spell(m_spell_name,
                                 m_spell_level,
                                 dummy))
                                   );

  // Try to have the player learn this spell, this can throw
  player.learn(*spell);
}

///////////////////////////////////////////////////////////////////////////////
std::string LearnCommand::help(const Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  const TalentTree& talents = engine.player().talents();

  std::vector<std::pair<std::string, unsigned> > learnable_spells;
  talents.query_all_learnable_spells(learnable_spells, engine.player());

  std::string help_str =
"<spell-name> <level>\n"
"  Player learns spell of type <spell-name> and level <level>\n"
"  Learnable spells:\n";
  for (std::vector<std::pair<std::string, unsigned> >::const_iterator
       itr = learnable_spells.begin(); itr != learnable_spells.end(); ++itr) {
    std::string spell_name = itr->first;
    unsigned spell_lvl     = itr->second;
    std::ostringstream out;
    out << "    " << spell_name << " : " << spell_lvl << "\n";
    help_str += out.str();
  }

  return create_help_str(this, help_str);
}

/*****************************************************************************/

///////////////////////////////////////////////////////////////////////////////
void DrawCommand::init(const std::vector<std::string>& args)
///////////////////////////////////////////////////////////////////////////////
{
  RequireUser(args.size() == 1, "The draw command takes 1 argument");

  // Parse draw mode
  s_draw_mode = args[0];
}

///////////////////////////////////////////////////////////////////////////////
void DrawCommand::apply(Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  DrawMode new_draw_mode = Drawable::parse_draw_mode(s_draw_mode);
  Drawable::set_draw_mode(new_draw_mode);
  engine.interface().draw(); // redraw
}

///////////////////////////////////////////////////////////////////////////////
std::string DrawCommand::help(const Engine& engine) const
///////////////////////////////////////////////////////////////////////////////
{
  // TODO - Needs to be decoupled from exact contents of drawable enum. IE
  //        this help messages needs to auto-update when draw-modes change.
  std::string usage =
"<draw-mode>\n"
"  Changes how the world is drawn.\n"
"  Available draw modes:\n";
  for (DrawMode mode_itr = Drawable::FIRST; ; ++mode_itr) {
    usage += "    " + Drawable::draw_mode_to_str(mode_itr)  + "\n";
    if (mode_itr == Drawable::LAST) {
      break;
    }
  }

  return create_help_str(this, usage);
}

File: code/game/command.py
#! /usr/bin/env python

import time

from spell_factory import SpellFactory
from baal_common import prequire, urequire, Location, UserError, grant_access
from engine import engine
from draw_mode import DrawMode, _set_draw_mode, explain_draw_mode
from player import Player

###############################################################################
class Command(object):
###############################################################################
    """
    A Command is the entity by which a player affects the game state.
    This class is an abstract base class intended to impose an interface on all
    Commands. Subclasses of Command are automatically added to the
    CommandFactory.
    """

    def __init__(self): prequire(False, "Do not instantiate Command")

    #
    # ==== Public API ====
    #

    #
    # Instance API
    #

    def apply(self):
        """
        Apply this command to this system.
        """
        prequire(False, "Called abstract version of apply")

    #
    # Class-method API (with default implementations)
    #

    @classmethod
    def name(cls):
        """
        Get the name of the command, this is what the user will type to issue
        the command.
        """
        return cls._NAME

    @classmethod
    def aliases(cls):
        """
        Get aliases for the command, the user can type these as a shortcut for
        the command.
        """
        return cls._ALIASES

    @classmethod
    def help(cls, extra_args=None):
        """
        Get help information for this command.
        """
        if (_is_text_interface()):
            return _create_text_help_str(cls, cls._TEXT_USAGE)
        else:
            return cls._GRAPHICAL_USAGE

    #
    # ==== Class Constants ====
    #

    # These should be overridden
    _NAME            = None
    _ALIASES         = None
    _TEXT_USAGE      = None
    _GRAPHICAL_USAGE = None

#
# Command implementations. The rest of the system does not care about anything
# below. Only the abstract base class (Command) should be exposed.
#

###############################################################################
class _HelpCommand(Command):
###############################################################################

    # Class variables
    _NAME            = "help"
    _ALIASES         = ("h",)
    _TEXT_USAGE      = \
"""%s [<command>]
  Returns info/syntax help for a command or all commands if no argument
""" % _NAME
    _GRAPHICAL_USAGE = "Right click on items to get help"

    ###########################################################################
    def __init__(self, args):
    ###########################################################################
        self.__cmd = None
        self.__extra_args = None
        if (args):
            # Verify arg is a valid command name and get handle to cmd class
            from command_factory import CommandFactory
            self.__cmd = CommandFactory.get(args[0])
            if (len(args) > 1):
                self.__extra_args = args[1:]

    ###########################################################################
    def apply(self):
    ###########################################################################
        interface = engine().interface()

        # If user provided arg, give help for that cmd, otherwise give help for
        # all commands.
        if (self.__cmd is not None):
            interface.help(self.__cmd.help(self.__extra_args))
        else:
            from command_factory import CommandFactory
            for cmd_cls in CommandFactory:
                interface.help(cmd_cls.help())

###############################################################################
class _EndTurnCommand(Command):
###############################################################################

    # Class variables
    _NAME             = "end"
    _ALIASES          = ("n",)
    _TEXT_USAGE       = \
"""%s [<num-turns>]
Ends the current turn. Optional arg to skip ahead many turns
""" % _NAME
    _GRAPHICAL_USAGE  = "Ends the current turn."
    __MAX_SKIP_TURNS   = 100

    ###########################################################################
    def __init__(self, args):
    ###########################################################################
        urequire(len(args) <= 1, self.name(), " takes at most one argument")

        self.__num_turns = 1
        if (args):
            # Parse num turns
            try:
                self.__num_turns = int(args[0])
            except ValueError:
                urequire(False, "'%s' is not a valid integer" % args[0])
            urequire(self.__num_turns > 0 and
                     self.__num_turns <= self.__MAX_SKIP_TURNS,
                     "num-turns must be between 0 and ", self.__MAX_SKIP_TURNS)

    ###########################################################################
    def apply(self):
    ###########################################################################
        engine().interface().end_turn(self.__num_turns)

###############################################################################
class _QuitCommand(Command):
###############################################################################

    # Class variables
    _NAME            = "quit"
    _ALIASES         = ("q", "exit")
    _TEXT_USAGE      = \
"""%s
  Ends the game.
""" % _NAME
    _GRAPHICAL_USAGE = "Ends the game."

    ###########################################################################
    def __init__(self, args):
    ###########################################################################
        urequire(len(args) == 0, self.name(), " takes no arguments")

    ###########################################################################
    def apply(self):
    ###########################################################################
        engine().interface().end_turn()
        engine().quit()

###############################################################################
class _SaveCommand(Command):
###############################################################################

    # Class variables
    _NAME            = "save"
    _ALIASES         = ("s",)
    _TEXT_USAGE      = \
"""%s [<filename>]
  Saves the game; if no name provided, a name based on data/time will be used.
""" % _NAME
    _GRAPHICAL_USAGE = "Save the game."

    ###########################################################################
    def __init__(self, args):
    ###########################################################################
        urequire(len(args) <= 1, self.name(), " takes at most one argument")

        if (args):
            self.__savegame = args[0]
        else:
            self.__savegame = time.strftime("baal_%Y-%m-%d__%H:%M:%S.save")

    ###########################################################################
    def apply(self):
    ###########################################################################
        # TODO: Aaron, please re-implement in python
        urequire(False, "Not implemented yet")

###############################################################################
class _CastCommand(Command):
###############################################################################

    # Class variables
    _NAME            = "cast"
    _ALIASES         = ("c", "spell")
    _TEXT_USAGE      = \
"""%s <spell-name> <level> <row,col>
  Casts spell of type <spell-name> and level <level> at location <row>,<col>
""" % _NAME
    _GRAPHICAL_USAGE = \
"""
Click to cast a specific spell. Once you've clicked this button, your next
click should be on the square of the world where you want the spell to be
casted.
"""

    ###########################################################################
    def __init__(self, args):
    ###########################################################################
        urequire(len(args) == 3, self.name(), " takes three arguments")

        # Get spell name
        self.__spell_name = args[0]

        # Parse spell level
        try:
            self.__spell_level = int(args[1])
        except ValueError:
            urequire(False, "arg ", "'%s' is not a valid integer" % args[1])

        # Parse location
        try:
            self.__spell_location = Location.parse(args[2])
        except UserError:
            urequire(False, "arg ", "'%s' not a valid location" % args[2])

        grant_access(self, Player.ALLOW_CAST)
        grant_access(self, Player.ALLOW_GAIN_EXP)

    ###########################################################################
    def apply(self):
    ###########################################################################
        world  = engine().world()
        player = engine().player()

        # Ensure location is in-bounds
        urequire(world.in_bounds(self.__spell_location),
                 "Location ", self.__spell_location, " out of bounds. ",
                 "Max row is: ", world.height() - 1,
                 ", max col is: ", world.width() - 1)

        # Create the spell.
        spell = SpellFactory.create_spell(self.__spell_name,
                                          self.__spell_level,
                                          self.__spell_location)

        # Verify that player can cast this spell (can throw)
        player.verify_cast(spell)

        # Verify that it makes sense to cast this exact spell (can throw)
        spell.verify_apply()

        # These last two operations need to be atomic, neither should
        # ever throw a user error.
        try:
            # Let the player object know that the spell has been cast
            # and to adjust it's state accordingly.
            player.cast(spell)

            # Apply the spell to the world
            exp = spell.apply()

            # Give player experience
            player.gain_exp(exp)
        except UserError, e:
            prequire(False, "User error interrupted atomic operations: ", e)

    ###########################################################################
    @classmethod
    def help(cls, extra_args=None):
    ###########################################################################
        if (_is_text_interface()):
            full_usage = "%s  Castable spells:\n" % cls._TEXT_USAGE

            # Add info on castable spells to usage string
            for spell_name, spell_level in engine().player().talents():
                full_usage += "    %s : %d\n" % (spell_name, spell_level)

            return _create_text_help_str(cls, full_usage)
        else:
            prequire(extra_args is not None and len(extra_args) == 1,
                     "Expect spell name in extra_args")

            spell_name = extra_args[0]
            spell = SpellFactory.create_spell(spell_name)

            help_str = cls._GRAPHICAL_USAGE
            help_str += "\nDescription of %s spell:\n" % spell_name
            help_str += spell.info() + "\n"
            help_str += "Player has skill level %d in this spell\n" % \
                engine().player().spell_skill(spell_name)

            return help_str

###############################################################################
class _LearnCommand(Command):
###############################################################################

    # Class variables
    _NAME            = "learn"
    _ALIASES         = ("l",)
    _TEXT_USAGE      = \
"""%s <spell-name>
  Player learns spell of type <spell-name>. If spell already known, then
  skill in this spell is increased by one.
""" % _NAME
    _GRAPHICAL_USAGE = \
"""
Click to learn this skill OR increase skill in this spell.
"""

    ###########################################################################
    def __init__(self, args):
    ###########################################################################
        urequire(len(args) == 1, self.name(), " takes one argument")

        self.__spell_cls = SpellFactory.get(args[0])

        grant_access(self, Player.ALLOW_LEARN)

    ###########################################################################
    def apply(self):
    ###########################################################################
        engine().player().learn(self.__spell_cls.name())

    ###########################################################################
    @classmethod
    def help(cls, extra_args=None):
    ###########################################################################
        if (_is_text_interface()):
            full_usage = "%s  Learnable spells:\n" % cls._TEXT_USAGE

            # Add info on learnable spells to usage string
            for spell_name, spell_level in engine().player().learnable():
                full_usage += "    %s%s\n" % (spell_name,
                                          "" if spell_level == 1 else " (new)")

            return _create_text_help_str(cls, full_usage)
        else:
            prequire(extra_args is not None and len(extra_args) == 1,
                     "Expect spell name in extra_args")

            spell_name = extra_args[0]
            spell = SpellFactory.create_spell(spell_name)

            help_str = cls._GRAPHICAL_USAGE
            help_str += "\nDescription of %s spell:\n" % spell_name
            help_str += spell.info() + "\n"
            help_str += "Player has skill level %d in this spell\n" % \
                engine().player().spell_skill(spell_name)

            return help_str

###############################################################################
class _DrawCommand(Command):
###############################################################################

    # Class variables
    _NAME            = "draw"
    _ALIASES         = ("d", "show")
    _TEXT_USAGE      = \
"""%s [<draw-mode>]
  Re-draw the world; if arg is provided, draw mode is switched to that mode.
  Available draw modes:
    %s
""" % (_NAME,
       "\n    ".join([str(mode) for mode in DrawMode]))
    _GRAPHICAL_USAGE = \
"""
Re-draw the world with a diverent overlay.
"""

    ###########################################################################
    def __init__(self, args):
    ###########################################################################
        urequire(len(args) <= 1, self.name(), " takes at most one argument")

        # Parse draw mode
        self.__new_mode = None
        if (args):
            self.__new_mode = DrawMode(args[0])

    ###########################################################################
    def apply(self):
    ###########################################################################
        if (self.__new_mode is not None):
            # DrawCommand has the right to change draw mode
            _set_draw_mode(self.__new_mode)

        engine().interface().draw() # redraw

    ###########################################################################
    @classmethod
    def help(cls, extra_args=None):
    ###########################################################################
        if (_is_text_interface()):
            return _create_text_help_str(cls, cls._TEXT_USAGE)
        else:
            prequire(extra_args is not None and len(extra_args) == 1,
                     "Expect draw_mode in extra_args")
            draw_mode = extra_args[0]

            help_str =  cls._GRAPHICAL_USAGE
            help_str += "\nDescription of draw-mode %s:\n" % draw_mode
            help_str += explain_draw_mode(draw_mode)

            return help_str

###############################################################################
class _HackCommand(Command):
###############################################################################

    # Class variables
    _NAME            = "hack"
    _ALIASES         = ()
    _TEXT_USAGE      = \
"""%s [<exp>]
  Gives the player free arbitrary exp. If no arg is provided, enough exp is
  awarded to get the player to the next level. This is a cheat put in for
  testing.
""" % _NAME
    _GRAPHICAL_USAGE = \
"""
Click to give yourself a free level!
"""

    ###########################################################################
    def __init__(self, args):
    ###########################################################################
        urequire(len(args) <= 1, self.name(), " takes at most one argument")

        # Parse exp
        self.__exp = None
        if (args):
            try:
                self.__exp = int(args[0])
            except ValueError:
                urequire(False, "'%s' is not a valid integer" % args[0])

    ###########################################################################
    def apply(self):
    ###########################################################################
        player = engine().player()

        if (self.__exp is not None):
            exp_gained = self.__exp
        else:
            exp_gained = player.next_level_cost() - player.exp()

        engine().player().gain_exp(exp_gained)

#
# Internal methods
#

###############################################################################
def _create_text_help_str(cmd_cls, usage):
###############################################################################
    help_str = usage

    aliases = cmd_cls.aliases()
    if (aliases):
        help_str += "  Aliases: " + " ".join(aliases)

    return help_str

###############################################################################
def _is_text_interface():
###############################################################################
    # Importing interface within the global namespace will cause a circular
    # import, so we must wrap is_text_interface within an internal free func
    from interface import is_text_interface
    return is_text_interface()

#
# Tests
#

###############################################################################
def _test():
###############################################################################
    pass

if (__name__ == "__main__"):
    _test()
